src/tools/wasm-interp.cc

```diff
@@ -161,33 +175,147 @@ Result RunAllExports(const Instance::Ptr& instance, Errors* errors) {
   return result;
 }
 
-static void BindImports(const Module::Ptr& module, RefVec& imports) {
+static bool IsHostPrint(const ImportDesc& import) {
+  return import.type.type->kind == ExternKind::Func &&
+         (s_host_print &&
+          import.type.module == "host" &&
+          import.type.name == "print");
+}
+
+#if WITH_WASI
+static bool IsWasiImport(const ImportDesc& import) {
+  return import.type.type->kind == ExternKind::Func && 
+         (import.type.module == "wasi_snapshot_preview1" ||
+          import.type.module == "wasi_unstable");
+}
+
+static bool HasWasiImport(const Module::Ptr module) {
+  for (auto&& import : module->desc().imports) {
+    if (IsWasiImport(import)) {
+      return true;
+    }
+  }
+  return false;
+}
+#endif
+
+static Ref GenerateHostPrint(const ImportDesc& import) {
   auto* stream = s_stdout_stream.get();
 
+  auto func_type = *cast<FuncType>(import.type.type.get());
+  auto import_name = StringPrintf("%s.%s", import.type.module.c_str(),
+                                  import.type.name.c_str());
+
+  auto host_func = HostFunc::New(
+      s_store, func_type,
+      [=](Thread& thread, const Values& params, Values& results,
+          Trap::Ptr* trap) -> Result {
+        WriteCall(stream, import_name, func_type, params, results, *trap);
+        return Result::Ok;
+      });
+
+  return host_func.ref();
+}
+
+static Extern::Ptr GetImport(const std::string& module,
+                             const std::string& name) {
+  auto mod_iter = s_registry.find(module);
+  if (mod_iter != s_registry.end()) {
+    auto extern_iter = mod_iter->second.find(name);
+    if (extern_iter != mod_iter->second.end()) {
+      return extern_iter->second;
+    }
+  }
+  return {};
+}
+
+static void PopulateExports(const Instance::Ptr& instance,
+                            Module::Ptr& module,
+                            ExportMap& map) {
+  map.clear();
+  // Module::Ptr module{s_store, instance->module()};
+  for (size_t i = 0; i < module->export_types().size(); ++i) {
+    const ExportType& export_type = module->export_types()[i];
+    auto export_ = s_store.UnsafeGet<Extern>(instance->exports()[i]);
+    map[export_type.name] = export_;
+  }
+}
+
+static std::string GetPathName(std::string module_arg) {
+  size_t path_at = module_arg.find_first_of(':');
+
+  path_at = path_at == std::string::npos ? 0 : path_at + 1;
+
+  return module_arg.substr(path_at);
+}
+
+static std::string FileNameOfPath(std::string path_name) {
```

Comment:
I think you can use `GetBasename` + `StripExtension` from `wabt/filenames.h`?

Reply:
Thanks for the notice. I did not even know that exists already.